File: src/main/java/com/sems/shared/errors/AppException.java
package com.sems.shared.errors;

/**
 * Error de aplicacion comun a todos los modulos.
 *
 * <p>Traduce el {@code AppError} que los servicios en Go compartian entre capas.
 * El dominio lanza esta excepcion y la capa REST la convierte en el mismo JSON
 * que ya consume el frontend: <code>{"code": "...", "message": "..."}</code>.
 *
 * <p>Mantener el contrato de error identico es lo que permite cambiar la
 * implementacion del backend sin tocar el manejo de errores del cliente.
 */
public class AppException extends RuntimeException {

    public enum Code {
        VALIDATION_ERROR,
        NOT_FOUND,
        CONFLICT,
        UNAUTHORIZED,
        INTERNAL_ERROR
    }

    private final Code code;

    public AppException(Code code, String message) {
        super(message);
        this.code = code;
    }

    public Code code() {
        return code;
    }

    public static AppException validation(String message) {
        return new AppException(Code.VALIDATION_ERROR, message);
    }

    public static AppException notFound(String message) {
        return new AppException(Code.NOT_FOUND, message);
    }

    public static AppException conflict(String message) {
        return new AppException(Code.CONFLICT, message);
    }

    public static AppException unauthorized(String message) {
        return new AppException(Code.UNAUTHORIZED, message);
    }

    public static AppException internal(String message) {
        return new AppException(Code.INTERNAL_ERROR, message);
    }
}

File: src/Sems.Api/Shared/Errors/AppException.cs
namespace Sems.Api.Shared.Errors;

/// <summary>Categorias de error que reconoce la aplicacion.</summary>
public enum ErrorCode
{
    VALIDATION_ERROR,
    NOT_FOUND,
    CONFLICT,
    UNAUTHORIZED,
    INTERNAL_ERROR
}

/// <summary>
/// Error de aplicacion comun a todos los modulos.
///
/// <para>El dominio lanza esta excepcion y la capa HTTP la convierte en el mismo
/// JSON que ya consume el frontend: <c>{"code": "...", "message": "..."}</c>.
/// Mantener el contrato de error identico es lo que permite cambiar la
/// implementacion del backend sin tocar el manejo de errores del cliente.</para>
/// </summary>
public sealed class AppException : Exception
{
    public ErrorCode Code { get; }

    public AppException(ErrorCode code, string message) : base(message) => Code = code;

    public static AppException Validation(string message) => new(ErrorCode.VALIDATION_ERROR, message);

    public static AppException NotFound(string message) => new(ErrorCode.NOT_FOUND, message);

    public static AppException Conflict(string message) => new(ErrorCode.CONFLICT, message);

    public static AppException Unauthorized(string message) => new(ErrorCode.UNAUTHORIZED, message);

    public static AppException Internal(string message) => new(ErrorCode.INTERNAL_ERROR, message);
}

/// <summary>Cuerpo de error devuelto por la API.</summary>
public sealed record ErrorResponse(string Code, string Message);
